noise_corrector.py

```diff
@@ -1,66 +1,211 @@
-  outputs = get_possible_event_source_outputs()
-  true_reports = [
-      100_000_000,  # Sources with no reports
-      100_000,  # Sources with trigger (conversion) data "0"
-      40_000,  # Sources with trigger (conversion) data "1"
-  ]
-  # Simulate what the noisy reports produced by the browser would look like
-  noisy_reports = simulate_randomization(
-      true_reports, RANDOMIZED_RESPONSE_RATE_EVENT_SOURCES)
-  # Generate the noise-corrected reports 
-  # To use this script in a real system, replace `noisy_reports` with the actual attribution reports received by the endpoint (and sent from users' browsers).
-  # Note how `noisy_reports` is NOT a list of reports; it's an array of counts of reports of each kind
-  # For example, for view reports: `noisy_reports = [42, 10, 13]` would mean that among the 147 views that you recorded: 42 of them resulted in no report, 10 of them resulted in a report with trigger data 0, and 13 of them in a report with trigger data 1
-  corrected_reports = correct_event_sources(noisy_reports)
-
-  # Print the result in a tabular form
-  column_names = ["Output", "True count", "Noisy count", "Corrected count"]
-  print("{:<30}{:<20}{:<20}{:<20}".format(*column_names))
-  for i, v in enumerate(corrected_reports):
-    d = json.dumps(outputs[i])
-    print(f"{d:<30}{true_reports[i]:<20,}{noisy_reports[i]:<20,}{v:<20,.1f}")
+################## GENERIC UTILITIES ##################
+
+# Useful type declarations for working with JSON:
+class SourceRegistrationConfig(TypedDict, total=False):
+  source_event_id: str
+  expiry: int # Optional
+
+
+class Source(TypedDict):
+  source_time: int
+  source_type: str
+  registration_config: SourceRegistrationConfig
+
+
+class ReportRaw(TypedDict):
+  source_event_id: str
+  source_type: str
+  trigger_data: str
+
+
+class Report(TypedDict):
+  report_time: int
+  report: ReportRaw
+
+
+Joined = Tuple[Source, List[Report]]
+JoinedList = List[Joined]
+JoinedGen  = Generator[Joined, None, None]
+# Raw output configurations for correcting randomized response.
+# First int is the window index, followed by the trigger data.
+OutputConfig = Tuple[Tuple[int, int], ...]
+
+# For generic functions.
+T = TypeVar('T')
+
+
+class ParamConfig:
+  def __init__(self,
+               randomized_reponse_rate,
+               trigger_data_cardinality,
+               max_reports,
+               max_windows,
+               name):
+    self.rr_rate = randomized_reponse_rate
+    self.data_cardinality = trigger_data_cardinality
+    self.max_reports = max_reports
+    self.max_windows = max_windows
+    self.name = name
+
+
+NAV_PARAMS = ParamConfig(.0024, 8, 3, 3, 'navigation')
+EVENT_PARAMS = ParamConfig(.0000025, 2, 1, 1, 'event')
+
+DEFAULT_EXPIRY = 60 * 60 * 24 * 30
+
+class OutputEnumeration:
+  # Class encapsulating logic which represents the logical output of the API
+  # per source event, for the purposes of debiasing the randomized response
+  # privacy mechanism.
```

Comment:
Nit for clarity:
```suggestion
  # Class encapsulating logic which represents all possible outputs of the API
  # per source event, taking into account the randomized response
  # privacy mechanism.
```
(is it?)

Reply:
Gotcha, thank you for clarifying; whichever you think works best (the "enumeration" term was indeed why I was confused)